packages/application-shell/src/components/navbar/navbar-new/navbar.tsx

```diff
@@ -103,57 +104,61 @@ const getIsSubmenuRouteActive = (
-  const menuItemRef = useRef<HTMLDivElement>(null);
   const submenuRef = useRef<HTMLUListElement>(null);
 
   const hasSubmenu =
     Array.isArray(props.menu.submenu) && props.menu.submenu.length > 0;
 
-  const menuItemBoundingClientRect =
-    menuItemRef.current?.getBoundingClientRect();
-  const menuItemTop = menuItemBoundingClientRect?.top || 0;
-  const menuItemBottom = menuItemBoundingClientRect?.bottom || 0;
+  const menuItemIdentifier = snakeCase(props.menu.key);
+
+  const callbackFn: IntersectionObserverCallback = useCallback((entries) => {
+    const menuItemBoundingClientRect = document
+      .querySelector(`[data-menuitem="${menuItemIdentifier}"]`)
+      ?.getBoundingClientRect();
+    const menuItemTop = menuItemBoundingClientRect?.top || 0;
+    const menuItemBottom = menuItemBoundingClientRect?.bottom || 0;
+
+    const [entry] = entries;
+
+    const doesSubmenuFitWithinViewportBelowMenuItem =
+      entry.boundingClientRect.height +
+        (props.isMenuOpen ? menuItemTop : menuItemBottom) >
+      window.innerHeight;
+    // if the submenu does not fit at the bottom of the viewport (below the menu item)
+    if (doesSubmenuFitWithinViewportBelowMenuItem) {
+      setIsSubmenuAboveMenuItem(true);
+      setSubmenuVerticalPosition(
+        window.innerHeight - (props.isMenuOpen ? menuItemBottom : menuItemTop)
+      );
+      // show the submenu above the menu item
+    } else {
+      setIsSubmenuAboveMenuItem(false);
+      setSubmenuVerticalPosition(
+        props.isMenuOpen ? menuItemTop : menuItemBottom
+      );
+    }
+  }, []);
+
+  const observerRef = useRef<IntersectionObserver | null>(null);
 
-  const callbackFn: IntersectionObserverCallback = useCallback(
-    (entries) => {
-      const [entry] = entries;
-      const doesSubmenuFitWithinViewportBelowMenuItem =
-        entry.boundingClientRect.height +
-          (props.isMenuOpen ? menuItemTop : menuItemBottom) >
-        window.innerHeight;
-      // if the submenu does not fit at the bottom of the viewport (below the menu item)
-      if (doesSubmenuFitWithinViewportBelowMenuItem) {
-        setIsSubmenuAboveMenuItem(true);
-        setSubmenuVerticalPosition(
-          window.innerHeight - (props.isMenuOpen ? menuItemBottom : menuItemTop)
-        );
-        // show the submenu above the menu item
-      } else {
-        setIsSubmenuAboveMenuItem(false);
-        setSubmenuVerticalPosition(
-          props.isMenuOpen ? menuItemTop : menuItemBottom
-        );
-      }
-    },
-    [menuItemBottom, menuItemTop, props.isMenuOpen]
-  );
   useLayoutEffect(() => {
-    const observer = new IntersectionObserver(callbackFn, {
+    observerRef.current = new IntersectionObserver(callbackFn, {
       rootMargin: '-100% 0px 0px 0px', // we want to observe if the submenu crosses the bottom line of the viewport - therefore we set the root element top margin to -100% of the viewport height
     });
+    return () => observerRef.current?.disconnect();
+  }, []);
```

Comment:
The main idea behind this PR is to cache the `IntersectionObserver` instance and create it only at the first render. This idea was brought up [here](https://github.com/commercetools/merchant-center-application-kit/pull/3204#discussion_r1328477809).